src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs

```diff
@@ -38,96 +36,76 @@ public interface IConcurrencyManager
-
          var results = new ConcurrentDictionary<TData, ConcurrencyManagerResult<TResult>>();
 
-         //Splits the action based in the number of cores available. 
-         //No thread will be created here. If the actions are all running on the same thread, the effect of the concurrency execution will be inexistent
-
-         var tasks = Enumerable.Range(0, numberOfCoresToUse).Select(async coreId =>
-         {
-            //While there is data left
-            while (concurrentData.TryDequeue(out var datum))
-            {
-               cancellationToken.ThrowIfCancellationRequested();
-
-               //Invoke the action on it and store the result. We assume here that each action runs on its own thread.
-               var result = await returnWithExceptionHandling(
-                  coreId,
-                  action,
-                  datum,
-                  idFunc,
-                  cancellationToken
-               );
-               results.TryAdd(datum, result);
-            }
-         });
-
-         await Task.WhenAll(tasks);
-
-         //all tasks are completed. Can return results
+         await Task.Run(() => Parallel.ForEach(data, createParallelOptions(cancellationToken),
+               datum =>
+               {
+                  cancellationToken.ThrowIfCancellationRequested();
+                  try
+                  {
+                     results.TryAdd(
+                        datum, 
+                        new ConcurrencyManagerResult<TResult>(datum.Id, action(datum, cancellationToken))
+                     );
+                  }
+                  catch (Exception e)
+                  {
+                     results.TryAdd(
+                        datum,
+                        new ConcurrencyManagerResult<TResult>(datum.Id, e.Message)
+                     );
+                  }
+                  
+               }), cancellationToken);
          return results;
       }
 
-      private async Task<ConcurrencyManagerResult<TResult>> returnWithExceptionHandling<TData, TResult>
-      (
-         int coreId,
-         Func<int, TData, CancellationToken, Task<TResult>> task,
-         TData data,
-         Func<TData, string> idFunc,
-         CancellationToken cancellationToken
-      )
+      private ParallelOptions createParallelOptions(CancellationToken token)
       {
-         try
-         {
-            return new ConcurrencyManagerResult<TResult>
-            (
-               id: idFunc(data),
-               result: await task(coreId, data, cancellationToken)
-            );
-         }
-         catch (Exception e)
+         return new ParallelOptions()
          {
-            return new ConcurrencyManagerResult<TResult>
-            (
-               id: idFunc(data),
-               errorMessage: e.ExceptionMessage(addContactSupportInfo: false)
-            );
-         }
+            CancellationToken = token,
+            MaxDegreeOfParallelism = _maximumNumberOfCoresToUse
```

Comment:
@msevestre No, the idea was to have the ability to overwrite the settings. So if you say for example the maximum number of cores in settings is 100 but for some reason you want to use less on a single call, you can still pass 50 on the function call. If you pass 0 (default) or lower it will use the value from settings. We could remove this argument though if you think it is overkilling. The settings can be change any time anyways so you could theoretically change the value in the settings, call the function and then restore the settings.

Reply:
Exactly. That's why using `_maximumNumberOfCoresToUse` is wrong because it's always this value and is set in the constructor
```
 _maximumNumberOfCoresToUse = Math.Max(1, coreUserSettings.MaximumNumberOfCoresToUse);

```
I think it should be numberOfCoreToUse or sthg like this

in fact the value `numberOfCoresToUse ` is only set but never used. Something is clearly off

Reply:
and while we are at it, numberOfCoresToUse =0 as default is very, very strange (or use a negative value??!?). Set it as null by default which would mean use the settings. Otherwise use this value no?

Reply:
I think it's good to have this parameter however (probably for testing scenarios et.c.)

Reply:
You are right. It was used before in the function createParallelOptions but at some point was removed lol. I also changed it to uint so you are not allowed to pass negative values (which are meaningless in this context indeed)